[TestFramework] Add README note about sharing fields and tests (#17663)
Azure/azure-sdk-for-net: sdk/core/Azure.Core.TestFramework/README.md

The `Download-DevOpsRecordings.ps1` would wait for active runs to finish before retrieving artifacts unless `-NoWait` switch is used.

**NOTE:** these scripts require being signed in with Azure CLI (https://docs.microsoft.com/cli/azure/authenticate-azure-cli?view=azure-cli-latest) and access to the internal DevOps project (https://dev.azure.com/azure-sdk/internal/) 

> ## Note on private/non-virtual fields in your clients (such as _clientDiagnostics) and InternalsVisibleTo

Reviewer: I think the problem here is not fields it's that `internal virtual Response<SparkSession> GetSparkSession` is internal virtual and is not forwarded to the base implementation.

That's why the fields are null.

Reviewer: In theory, you shouldn't be having/testing internal client methods.

Author: Hey @pakrym - See [my branch](https://github.com/chamons/azure-sdk-for-net/blob/d8f01a29df29d33dec94b76ccee678b7839c098b/sdk/synapse/Azure.Analytics.Synapse.Spark/src/Customization/SparkSessionClient.cs#L196) for how I took a first stab at implementing a LRO.

- I wanted to "hide" the versions which returned a session that was not ready and provide versions that returned a LRO to wait on, so I copy pasted the generated code into the customization and made it internal, then provided a public that wrapped.
- That code now in the customization needed to access _clientDiagnostics, which was null in this one test case.

I patched that branch together by looking at what other SDKs _are_ doing, which is not necessarily the same as what I _should_ be doing. 

Advice welcome here.

Author: I decided to make SparkSessionOperation/SparkStatementOperation take an ClientDiagnostics argument to remove the need for this property.

Reviewer 2: Yeah, looks like @pakrym is correct. The diagnostic scope interceptor is not doing its thing because this method is not public. If you make it public temporarily, you will see that there are other errors about not using the correct scope name.
If this operation is exposed only via the LRO, do we need the separate scope here or can we just use the LRO scope?

Reviewer 2: Added some questions and comments, thanks!

The `Download-DevOpsRecordings.ps1` would wait for active runs to finish before retrieving artifacts unless `-NoWait` switch is used.

**NOTE:** these scripts require being signed in with Azure CLI (https://docs.microsoft.com/cli/azure/authenticate-azure-cli?view=azure-cli-latest) and access to the internal DevOps project (https://dev.azure.com/azure-sdk/internal/) 

> ## Note on sharing fields between Generated and Customized code (such as _clientDiagnostics)

Reviewer 3: This is more about private fields in general than the code gen split.  If you added your own by hand, they'd have the same problem which makes it even more confusing.  :smile:

```suggestion
## Note on private/non-virtual fields in your clients (such as _clientDiagnostics)
```

Some bindings require code on the customized side to access fields that are generated. For example:

```csharp
    // Generated\SparkSessionClient.cs
    public partial class SparkSessionClient
    {

>         private readonly ClientDiagnostics _clientDiagnostics;

Reviewer 4: (extreme) nit:  You may want to close the brackets for these.   The unbalancing is making me twitch.... but that my just be me.   😄

```

```csharp
    // Customization\SparkSessionClient.cs
    internal virtual Response<SparkSession> GetSparkSession(int sessionId, bool? detailed = null, CancellationToken cancellationToken = default)
    {

>         using var scope = ClientDiagnostics.CreateScope("SparkSessionClient.GetSparkSession");

Reviewer 3: Should probably change this back to the original for it to make sense

```suggestion
        using var scope = _clientDiagnostics.CreateScope("SparkSessionClient.GetSparkSession");
```

```

> The direct approach of using _clientDiagnostics here will fail on tests that use the test framework. The proxy generated at runtime uses the constructor with no parameters and _clientDiagnostics will be null.

Reviewer 4: ```suggestion
The direct approach of using `_clientDiagnostics` here will fail on tests that use the test framework. The proxy generated at runtime uses the constructor with no parameters and _clientDiagnostics will be `null`.
```

> The correct solution is the use a virtual method, so the testing framework stubs it out:

Reviewer 3: ```suggestion
The correct solution is the use a virtual property, so the testing framework stubs it out:
```

> The correct solution is the use a virtual method, so the testing framework stubs it out:

Reviewer 4: ```suggestion
The correct solution is to use a virtual method, so the testing framework stubs it out:
```

```csharp
    internal virtual ClientDiagnostics ClientDiagnostics => _clientDiagnostics;
```

> This may require an InternalsVisisbleTo addition to the test AssemblyInfo.cs:

Reviewer 3: ```suggestion
Then update your client code to use the new virtual property.  (Pro tip: you can do a VS rename on the field from `_clientDiagnostics` to `ClientDiagnostics` and then change just the field/property definitions back to `_clientDiagnostics` to make this quicker.)

If it's an `internal` property, it will also require an `InternalsVisisbleTo` addition in your AssemblyInfo.cs:
```

You added it to the main AssemblyInfo, not the test one, right?  I thought that was required though it has been awhile since I've done this.

> This may require an InternalsVisisbleTo addition to the test AssemblyInfo.cs:

Reviewer 4: ```suggestion
This may require an `InternalsVisisbleTo` addition to the test `AssemblyInfo.cs`:
```

        private readonly ClientDiagnostics _clientDiagnostics;
        ...
    }
```

```csharp
    // Customization\SparkSessionClient.cs
    internal virtual Response<SparkSession> GetSparkSession(int sessionId, bool? detailed = null, CancellationToken cancellationToken = default)
    {
        using var scope = _clientDiagnostics.CreateScope("SparkSessionClient.GetSparkSession");
        ...
    }
```

The direct approach of using `_clientDiagnostics` here will fail on tests that use the test framework. The proxy generated at runtime uses the constructor with no parameters and _clientDiagnostics will be `null`.

The correct solution is to use a virtual property, so the testing framework stubs it out:

```csharp
    internal virtual ClientDiagnostics ClientDiagnostics => _clientDiagnostics;
```

Then update your client code to use the new virtual property.  (Pro tip: you can do a VS rename on the field from `_clientDiagnostics` to `ClientDiagnostics` and then change just the field/property definitions back to `_clientDiagnostics` to make this quicker.)

If it's an `internal` property, it will also require an `InternalsVisisbleTo` addition in your `AssemblyInfo.cs`:

```csharp

> [assembly: InternalsVisibleTo("DynamicProxyGenAssembly2, PublicKey=0024000004800000940000000602000000240000525341310004000001000100c547cac37abd99c8db225ef2f6c8a3602f3b3606cc9891605d02baa56104f4cfc0734aa39b93bf7852f7d9266654753cc297e7d2edfe0bac1cdcf9f717241550e0a7b191195b7667bb4f64bcb8e2121380fd1d9d46ad2d92d2d15605093924cceaf74c4861eff62abf69b9291ed0a340e113be11e6a7d3113e92484cf7045cc7")]

Reviewer 2: What do folks think about adding this to the template: https://github.com/Azure/azure-sdk-for-net/blob/master/sdk/template/Azure.Template/src/properties/AssemblyInfo.cs
